src/main/java/eu/dissco/core/digitalspecimenprocessor/service/ProcessingService.java

```diff
@@ -1,632 +1,417 @@
+    var existingPidMap = existingMedias.entrySet()
+        .stream().collect(toMap(
+            Entry::getKey,
+            e -> e.getValue().id()
+        ));
+    return concatMaps(existingPidMap, newMediaPids);
   }
 
-  private void gatherDigitalMediaObjectForEqualRecords(
-      List<DigitalSpecimenRecord> currentDigitalSpecimen, List<DigitalSpecimenEvent> events) {
-    var pidMap = currentDigitalSpecimen.stream().collect(Collectors.toMap(
-        digitalSpecimenRecord -> digitalSpecimenRecord.digitalSpecimenWrapper()
-            .physicalSpecimenID(),
-        DigitalSpecimenRecord::id
+  private static Map<String, String> concatMaps(Map<String, String> m1, Map<String, String> m2) {
+    return Stream.concat(
+        m1.entrySet().stream(), m2.entrySet().stream()
+    ).collect(toMap(
+        Entry::getKey,
+        Entry::getValue
     ));
-    // Events contain the incoming media information, the record contains the id. We need to link both
-    // to publish media events
-    for (var event : events) {
-      if (pidMap.containsKey(event.digitalSpecimenWrapper().physicalSpecimenID())) {
-        var digitalSpecimenPid = pidMap.get(event.digitalSpecimenWrapper().physicalSpecimenID());
-        var digitalMedia = event.digitalMediaEvents();
-        publishDigitalMediaRecord(digitalMedia, digitalSpecimenPid, null);
-      }
-    }
   }
 
-  private Set<DigitalSpecimenRecord> updateExistingDigitalSpecimen(
-      List<UpdatedDigitalSpecimenTuple> updatedDigitalSpecimenTuples) {
-    log.info("Persisting to DOI Server");
-    var successfullyUpdatedHandles = updateHandles(updatedDigitalSpecimenTuples);
-    if (!successfullyUpdatedHandles) {
-      return Set.of();
+  private List<DigitalSpecimenRecord> processSpecimenResults(
+      SpecimenProcessResult specimenProcessResult,
+      Map<String, PidProcessResult> pidProcessResults) {
+    var results = new ArrayList<DigitalSpecimenRecord>();
+    if (!specimenProcessResult.equalSpecimens().isEmpty()) {
+      digitalSpecimenService.updateEqualSpecimen(specimenProcessResult.equalSpecimens());
     }
-    var mediaPidMap = createMediaPidsForUpdatedRecords(
-        updatedDigitalSpecimenTuples);
-    var digitalSpecimenRecords = getSpecimenRecordMap(updatedDigitalSpecimenTuples, mediaPidMap);
-    log.info("Persisting {} updated record to the database", digitalSpecimenRecords.size());
-    try {
-      repository.createDigitalSpecimenRecord(
-          digitalSpecimenRecords.stream().map(UpdatedDigitalSpecimenRecord::digitalSpecimenRecord)
-              .toList());
-    } catch (DataAccessException e) {
-      log.error("Unable to update records into database. Rolling back updates", e);
-      rollbackService.rollbackUpdatedSpecimens(digitalSpecimenRecords, mediaPidMap, false, false);
-      return Collections.emptySet();
+    if (!specimenProcessResult.newSpecimens().isEmpty() && !specimenProcessResult.newSpecimenPids()
```

Comment:
What happens when the one is empty?

Reply:
Good catch. If the newSpecimenPids is empty, that means PIDs weren't created for the new speicmens, so we shouldn't create new speicmens or risk a NPE. Added a warning message to clarify things.